File: 3093-sum-of-values-at-indices-with-k-set-bits/sum-of-values-at-indices-with-k-set-bits.py
from typing import List

class Solution:
    def count_set_bits(self, num):
        count = 0
        while num:
            count += num & 1
            num >>= 1
        return count

    def sumIndicesWithKSetBits(self, nums: List[int], k: int) -> int:
        result = 0
        for i, num in enumerate(nums):
            if self.count_set_bits(i) == k:
                result += num
        return result

File: 3093-sum-of-values-at-indices-with-k-set-bits/sum-of-values-at-indices-with-k-set-bits.cpp
class Solution {
public:
    int sumIndicesWithKSetBits(vector<int>& nums, int k) {
        int n = nums.size();
        int sum = 0;
        
        // Precompute the number of set bits for all possible indices
        vector<int> bitCounts(1 << 20, 0); // Assuming nums.size() <= 2^20 for practicality
        for (int i = 1; i < (1 << 20); ++i) {
            bitCounts[i] = bitCounts[i >> 1] + (i & 1);
        }
        
        // Iterate through all indices and sum the values where the index has k set bits
        for (int i = 0; i < n; ++i) {
            if (bitCounts[i] == k) {
                sum += nums[i];
            }
        }
        
        return sum;
    }
};
